src/octoprint/settings.py

```diff
@@ -114,6 +114,7 @@ def settings(init=False, basedir=None, configfile=None):
 		"additionalPorts": [],
 		"additionalBaudrates": [],
 		"longRunningCommands": ["G4", "G28", "G29", "G30", "G32", "M400", "M226", "M600"],
+		"synchronousCommands": ["M109","M190","G4", "G28", "G29", "G30", "G32", "M400", "M226", "M600"],
```

Comment:
i'm working only with marlin 1.1.9-bugfix
it have 3 buffer layers
1) RX_BUFFER_SIZE - 2048bytes. in my configuration, it is incoming serial buffer filled  from serial interface interrupts

2) BUFSIZE - 32 commands. it contain unparsed lines (simply Gcode line), but checksum already checked 

3) BLOCK_BUFFER_SIZE - 64 commands. it is planner buffered commands (actual buffered commands in terms of reprap )

both BUFSIZE and BLOCK_BUFFER_SIZE processing in main loop as follow:

```

void loop() {
  void get_available_commands() {
    inline void get_serial_commands() {
      while (commands_in_queue < BUFSIZE && (c = MYSERIAL0.read()) >= 0) {
      }
  }
  process_next_command(){
    char * const current_command = command_queue[cmd_queue_index_r];
    parser.parse(current_command);
    process_parsed_command(){
               buffer_line() 
               planar.buffer_segment() {
                      block_buffer[BLOCK_BUFFER_SIZE]
                  }
      ok_to_send();
    }
  }
  --commands_in_queue;

  endstops.event_handler();
  idle(){
    //lcd_update();
    host_keepalive();
    thermalManager.manage_heater();
    thermalManager.auto_report_temperatures();
  }
}

```

"synchronous commands" is designed to prevent overflow in RX_BUFFER_SIZE, just clear this list and try to print, very quickly you will get resend request

as alternative we could send asynchronously only reprap "buffered commands"

Reply:
@foosel hi!
synchronousCommands - is a commands on which marlin may not send "ok" immediately, it will send busy or wait but after some timeout

imagine we have following gcode file
```
M80 ; power on
M190 S110
M109 S270
G1 ....
>100 lines of G0-G1 commands, each line is about 86 bytes long
```
after M80 we will get `ok P63 B31`
so marlin say that it ready to receive up to 94 commands, but
if we will send 94 lines from gcode file marlin will not accept them, instead it will stuck at M190 and will waiting for heating complete. at the same time in that moment it will not send to us nor ok nor busy nor wait, it will send busy later after timeout.

also, because of restricted RX_BUFFER_SIZE, when we send 94 * 86 = 8084 bytes RX_BUFFER will be overflowed, some command will be corrupted and marlin will send resend request, which will worsen print quality. 

so after send `M190 S110` we will not have any feedback to continue send commands asynchronously.

I think this is bug in marlin protocol but right now we can't resolve it, that is why we need special list with such commands no which marlin will not answer immediately.

this list is pretty close to `longRunningCommands`, i don't know could we merge this two lists or probably better create another one - `bufferedCommans` (instead synchronousCommands) which will contain buffered  commands according to reprap documentation, and send asynchronously only commands from it but as soon as send one command not from bufferedCommands stop sending asynchronously and continue sending synchronously.

Reply:
>Buffered commands should usually only be G0-G3 and G28-G32 (see also here).
i disagree, as i describe  earlier G28-G32 may not answer immediately.

also   following commands working well asynchronously
```
M420 S1 Z1.0 ; Load auto bed leveling data from epprom
M900 K0.025; titan aero 150
M900 S0 ; activate
M203 X500.00 Y500.00 Z5.00 E25.00 ;Max Feedrate
M204 P2000 R1000 T3000 ;Default Acceleration
M205  J0.1 E15
M201 X5000 Y5000 E5000 ;Max Acceleration

G92 E0 ; Reset Extruder
```

so list of bufferedCommands may be much longer than my current `synchronousCommands`

Reply:
> i disagree, as i describe earlier G28-G32 may not answer immediately.

No buffered command will answer immediately if the planner buffer is full. And I was going off of what I found across firmware.



> `bufferedCommands` in my understanding is synonym of asynchronous commands, correct me if i'm wrong

As far as I understand your interpretation of "asynchronous commands", yes.

I'd got with a list of buffered/asynchronous commands instead of the opposite since it's usually the shorter one (every command by default is unbuffered/synchronous unless it explicitly touches the planner buffer) and more likely to match across firmware variants. Remember that Marlin 1.1.9-bugfix is but a droplet in the ocean of firmware variants out there.



> as alternative we could send asynchronously only reprap "buffered commands"

I'd prefer that due to the aforementioned reason since it limits the risk of considering a command capable of this that in fact isn't on some firmware variant out there. And it also makes more sense from a "matching up with firmware implementations" point of view, since we'd stick to the planner buffer related commands that way instead of a seemingly arbitrary list that no one understands how it came to be.

Reply:
> I'd prefer that due to the aforementioned reason since it limits the risk of considering a command capable of this that in fact isn't on some firmware variant out there.

@foosel  done in last commit, working but needs for more testing

Reply:
The list of buffered commands still seems way too long. `M104` is definitely not buffered. It might be that some of the other ones are buffered on your specific fork, but not in general. We should really stick to universally buffered commands here which - again - are `G0` through `G3` and `G28` through `G32`. Anything more than that and we risk issues with firmware variants.